Add data model support to HAFS workflow

evankalina opened the pull request:
## Description of changes  
This PR allows developers to perform HAFS runs with the [Community Data models for Earth Prediction Systems (CDEPS)](https://escomp.github.io/CDEPS/html/index.html). Developers can use the new workflow capabilities to one-way couple the HYCOM ocean model to a data atmosphere created from the ERA5 reanalysis. Or they can one-way couple the UFS weather model to a data ocean created from the OISST or GHRSST analyses. The workflow is capable of running all of the necessary steps in an automated fashion, including pre-processing the input data, generating the data model grids, and running the forecast. Scripts are provided to download ERA5, OISST, and GHRSST data in an offline fashion. This PR is not expected to have any impact on HAFS runs that do not use data models.

## Dependencies
There are no submodule dependencies, as the necessary updates have already been made to the support/HAFS branch of CDEPS in a previous commit ([PR # 4](https://github.com/hafs-community/CDEPS/pull/4)).

A user guide that describes how to run CDEPS in the workflow, along with suggestions for how to use custom data sources to generate the data models, is available [here](https://docs.google.com/document/d/1_UsLo96LAxCZjXHZe6qrr2WsR5W8PC3CGJxHod9qWKE). It would be good to include this information in a new chapter of the HAFS Developers guide.

## Contributors
Along with the author, @SamuelTrahanNOAA, @uturuncoglu, and @danrosen25 worked to add CDEPS support to the HAFS workflow. We would also like to thank @rsdunlapiv, @mvertens, and @ligiabernardet for helping guide the design decisions.

## Tests conducted
On Orion, the full workflow was tested, with 126-h forecasts of the ERA5 data atmosphere and both the OISST and GHRSST data oceans for several cycles of Hurricanes Dorian and Zeta. On Jet and Hera, the entire workflow was tested except for the forecast and post-processing steps. The forecast jobs have been sitting in the Jet/Hera queues for weeks without running. We anticipate that Orion will be the primary machine used for data model runs in HAFS, and we can always submit a new PR if any changes are needed to fix bugs on Jet and Hera. We anticipate only needing to make small changes, if any, on those platforms.

## Application-level regression test status  
Running the HAFS application-level regression tests is currently performed by code reviewers after the developer creates the initial PR. As regression tests are conducted, the testers should use the checklist below to indicate **successful** regression tests. You may add other tests as needed. If a test fails, do not check the box. Instead, describe the failure in the PR comments, noting the platform where the test failed.

- [ ] Jet
- [x] Hera
- [x] Orion
- [x] WCOSS Dell P3
- [x] WCOSS Cray

Let's start the HAFS application/workflow level regression tests as you helped previously:
@evankalina, @panll, @mrinalbiswas (Hera and Orion)
@JunghoonShin-NOAA (Jet, we probably will skip Jet since it is occupied by the HFIP real-time parallel experiments under reservation)
@LinZhu-NOAA (wcoss_dell_p3)
@BijuThomas-NOAA (wcoss_cray)
Also, @JohnSteffen-NOAA please test the three HAFS CDEPS tests on wcoss_dell_p3.

danrosen25 committed:
Add CMEPS (mediator) to hafs_forecast submodule component and build.

BinLiu-NOAA committed:
Script changes in order to support run a CMEPS build for hafs-hycom coupling,
even when CMEPS is disabled, according to the instructions from Dan R.

BinLiu-NOAA committed:
Enable atmospheric surface pressure forcing in hafs-hycom coupling (Dan R. and Hyun-Sook)

danrosen25 committed:
Update CMEPS submodule. Includes ocean fraction computation fix.

BinLiu-NOAA committed:
Merge branch 'feature/hafs_couplehycom' into hafs_couplehycom_cmeps

Commit:
Add support for running a CDEPS data atmosphere (datm) to the HAFS
workflow.

Commit:
update model to support data components with CDEPS

danrosen25 committed:
Merge branch 'feature/hafs_couplehycom' into 'feature/hafs_couplehycom_cmeps'

evankalina committed:
Do not use INPdir as the CDEPS input data directory, since this
variable is already used locally by multiple scripts in the workflow.
The new CDEPS input data directory is called DATMdir and is specific
to the CDEPS DATM.

Commit:
update model to enable data components with cmake build

danrosen25 committed:
Update the hafs_forecast.fd submodule after bringing in FV3 regional
grid corners, HYCOM_merged branch, ability to set NEMS connector
attributes through config, and CMEPS datm mask value fix.

danrosen25 committed:
Merge branch 'feature/hafs_couplehycom' with 'feature/hafs_couplehycom_cmeps'. Update submodule for hafs_forecast.fd. CMEPS has been synchronized with ESCOMP/CMEPS and NEMS updated to include PIO initialization.

evankalina committed:
Point to feature/hafs_couplehycom_cdeps of ufs-weather-model

evankalina committed:
Finish pointing to feature/hafs_couplehycom_cdeps of ufs-weather-model

evankalina committed:
Merge remote-tracking branch 'origin/develop' into feature/hafs_couplehycom_cdeps

evankalina committed:
Advance hafs_gsi, hafs_post, and hafs_utils to the latest versions
used by b/develop.

evankalina committed:
A few small updates for CDEPS DATM

evankalina committed:
Update to the most recent version of the feature/hafs_couplehycom_cdeps branch of UFS weather model.

evankalina committed:
Merge remote-tracking branch 'origin/develop' into feature/hafs_couplehycom_cdeps

evankalina committed:
CDEPS no longer needs the pio_in and X_modelio.nml files.

evankalina committed:
Bug fix to set the year at the beginning and end of the forecast properly in datm.streams.xml

evankalina committed:
From Sam Trahan (CIRES at NOAA/GSL, DTC): Add a download script for
ERA5 reanalysis data for the CDEPS DATM.

SamuelTrahanNOAA committed:
Generate DATM mesh in workflow if make_mesh_atm=yes

evankalina committed:
Update to the latest version of the feature/hafs_couplehycom_cdeps branch of ufs-weather-model

SamuelTrahanNOAA committed:
Add DATM input (ERA5) download job. Rename ush/hafs_era5_downloader.py -> ush/hafs_datm_download.py

evankalina committed:
Sync with the latest ufs-weather-model feature/hafs_couplehycom_cdeps branch.

evankalina committed:
Add the capability to run CDEPS DOCN to the workflow.

evankalina committed:
Add comment to nems.configure for CDEPS so that the user isn't confused by the additional options there.

SamuelTrahanNOAA committed:
.gitignore Rocoto db files and Emacs backup files

SamuelTrahanNOAA committed:
Download DOCN files and generate their mesh. Fix a bug in hafs launcher data model checks.

SamuelTrahanNOAA committed:
.gitignore system.conf, fix directory, and generated rocoto xml files

SamuelTrahanNOAA committed:
Add QUEUE_FORECAST for forecast job. Orion defaults to "urgent"

SamuelTrahanNOAA committed:
Bug fixes to get DOCN+FV3 to run.

SamuelTrahanNOAA committed:
Changes to get DATM+HYCOM working.

SamuelTrahanNOAA committed:
Two files missing from prior commit: split hafs_cdeps.conf into hafs_datm.conf and hafs_docn.conf

SamuelTrahanNOAA committed:
Download an extra day of data before and after storm. Without this, DOCN fails.

SamuelTrahanNOAA committed:
Download an extra day of data before and after storm. Without this, DOCN fails.

SamuelTrahanNOAA committed:
Download one extra day of DATM data

SamuelTrahanNOAA committed:
Give DATM data 1 day before and after simulation to avoid future problems.

evankalina committed:
Small refinements to the implementation of CDEPS in the HAFS workflow.
Includes setting taxMode and yearAlign correctly and making system_type
a dynamic variable in the workflow.

evankalina committed:
Merge remote-tracking branch 'origin/develop' into feature/hafs_couplehycom_cdeps

Conflicts:
	modulefiles/modulefile.hafs.run.orion
	parm/forecast/regional_hwrf/nems.configure.atm_ocn.tmp
	parm/hafs.conf
	rocoto/hafs_workflow.xml.in
	rocoto/sites/hera.ent
	rocoto/sites/kjet.ent
	rocoto/sites/orion.ent
	rocoto/sites/wcoss_cray.ent
	rocoto/sites/wcoss_dell_p3.ent
	rocoto/sites/xjet.ent
	rocoto/sites/xjet_hafsv0p1a.ent
	rocoto/sites/xjet_hafsv0p2a.ent
	scripts/exhafs_forecast.sh
	sorc/build_forecast.sh
	sorc/hafs_forecast.fd

evankalina committed:
Make sure that the atmospheric initialization system is disabled if
the hafs_datm.conf is used.

evankalina committed:
Bug fix: When run_datm=yes, make sure the forecast job does not
start before ocn_prep finishes.

evankalina committed:
Bug fix: Make sure INPUT/ directory is created before attempting to
copy files to it. Clean up hafs.conf.

SamuelTrahanNOAA committed:
Corrections to hafs_esmf_mesh.py: print reason for import failure and request python3 instead of python

uturuncoglu committed:
add cheyenne support to build

evankalina committed:
Update submodule pointers to the latest hashes of feature/hafs_couplehycom_cdeps (ufs-weather-model) and support/HAFS (GSI, post, and utils).

evankalina committed:
Merge branch 'feature/hafsv0.2_phase3_final' into feature/hafs_couplehycom_cdeps

Conflicts:
	rocoto/hafs_workflow.xml.in
	rocoto/sites/xjet_hafsv0p1a.ent
	sorc/hafs_forecast.fd

evankalina committed:
Update the HAFS-CDEPS workflow to use dxxx.streams files in text format,
instead of XML. This change is required by recent updates to the CDEPS
implementation in UFS weather model.

evankalina committed:
Fix incomplete merge in wcoss_cray.ent.

SamuelTrahanNOAA committed:
split DOCN scripts into three, for GHRSST, OISST, and RTOFS

SamuelTrahanNOAA committed:
Allow the choice between several DOCN model sources. Presently only OISST is confirmed to work though.

SamuelTrahanNOAA committed:
Merge branch 'feature/hafs_couplehycom_cdeps' of https://github.com/hafs-community/HAFS into feature/hafs_couplehycom_cdeps

SamuelTrahanNOAA committed:
Make DATM configurable via datm_source. Fix scripts so one can run without docn. Correct some comments and usage messages.

SamuelTrahanNOAA committed:
Merge branch 'feature/hafs_couplehycom_cdeps' of https://github.com/hafs-community/HAFS into feature/hafs_couplehycom_cdeps

SamuelTrahanNOAA committed:
Use a python installation that already has the needed modules installed

evankalina committed:
Merge branch 'develop' into feature/hafs_couplehycom_cdeps and advance
submodule hashes to the latest version of feature/hafs_couplehycom_cdeps
(ufs-weather-model) and support/HAFS (all others).

Conflicts:
	sorc/build_forecast.sh
	sorc/hafs_forecast.fd

SamuelTrahanNOAA committed:
DOCN GHRSST works on Orion

SamuelTrahanNOAA committed:
Missed a "-p" in mkdir for DOCNdir and DATMdir

SamuelTrahanNOAA committed:
link the new fix_cdeps directory

SamuelTrahanNOAA committed:
Use fix-file meshes by default for DOCN. (DATM not yet tested this way)

SamuelTrahanNOAA committed:
Point to pregenerated mesh name in datm mode

SamuelTrahanNOAA committed:
Correct DOCN/DATM mask usage logic and add forecast.print_esmf option

SamuelTrahanNOAA committed:
Remove all unneeded vars from ghrsst before spatial subsetting

SamuelTrahanNOAA committed:
Point to alternative python on Jet

SamuelTrahanNOAA committed:
Reduce runtime for data model prep jobs now that GHRSST prep is much faster

SamuelTrahanNOAA committed:
Update paths for jet in cronjob_hafs_docn.sh

SamuelTrahanNOAA committed:
Put DATM input in intercom/cdeps just like DOCN

SamuelTrahanNOAA committed:
Put Hera DOCN resources in rocoto/sites/hera.ent

SamuelTrahanNOAA committed:
Update modulefiles for hera

SamuelTrahanNOAA committed:
Correct the docn/datm prep resources for kjet and hera

SamuelTrahanNOAA committed:
Correct python paths in data model cron scripts

evankalina committed:
Merge branch 'develop' into feature/hafs_couplehycom_cdeps
and advance submodule hashes.

evankalina committed:
Remove the redundant mesh_ocn_in entry in the docn configuration files.

evankalina committed:
Point to the support/HAFS branch of ufs-weather-model

BinLiu-NOAA commented:
@evankalina, wondering if we can have a tag-up telecon to discuss how to move forward with this PR. For example, how to support all the HAFS supported platforms? Which python version to use? How to deal with the Data download and/or Data_prep related tasks? etc.
Also, it will also be helpful if you or @SamuelTrahanNOAA can give us an introduction presentation about the details of the Data model component related workflow elements/changes. Thanks! Bin

evankalina commented:
@BinLiu-NOAA Sounds good to me. I sent an invite for us to discuss these changes in more detail, and I can give an overview presentation of the changes.

BinLiu-NOAA commented:
> @BinLiu-NOAA Sounds good to me. I sent an invite for us to discuss these changes in more detail, and I can give an overview presentation of the changes.

Thanks a lot, @evankalina!

uturuncoglu commented on parm/cdeps/datm_era5.streams:
```diff
@@ -0,0 +1,15 @@
+stream_info:               ERA5_HOURLY01
+taxmode01:                 limit 
+mapalgo01:                 bilinear
```
We might also test `mapalgo01: redist`. It think there is no need to use bilinear in here since stream and datm component uses same grid/mesh.

evankalina replied:
This change has been tested and incorporated into the PR.

uturuncoglu commented on parm/cdeps/docn_ghrsst.streams:
```diff
@@ -0,0 +1,15 @@
+stream_info:               PRESCRIBED01
+taxmode01:                 limit
+mapalgo01:                 bilinear
```
same also for this one, `mapalgo01:redist`

evankalina replied:
This change has been tested and incorporated into the PR.

uturuncoglu commented on parm/cdeps/docn_oisst.streams:
```diff
@@ -0,0 +1,15 @@
+stream_info:               PRESCRIBED01
+taxmode01:                 limit
+mapalgo01:                 bilinear
```
same also for this one, `mapalgo01:redist`

evankalina replied:
This change has been tested and incorporated into the PR.

uturuncoglu commented on parm/forecast/regional/nems.configure.cdeps.tmp:
```diff
@@ -0,0 +1,150 @@
+  start_type = startup
+::
+
+ALLCOMP_attributes::
```
Just for your information, the new version of CMEPS will also require setting following options under `ALLCOMP_attributes::`.

```
  stop_n = [forecast length need to be defined in here and need to be consistent with the model_configure]
  stop_option = nhours
  stop_ymd = -999
```

plus, new version of ESMF. So, once CMEPS is updated under HAFS (or UFS) ESMF module needs to be updated (8.2.0b?)

evankalina replied:
Thanks for letting us know! We can make these changes once CMEPS is updated in HAFS.

uturuncoglu commented on parm/hafs_basic.conf:
```diff
@@ -47,8 +47,15 @@ run_enkf=no          ;; Run self-cycled ensemble data assimilation system with E
 ocean_model=hycom
+run_datm=no          ;; Data atmosphere using CDEPS
+run_docn=no          ;; Data ocean using CDEPS
+run_dwav=no          ;; Data waves using CDEPS
```
I think there is no data wave capability at this point. Right?

evankalina replied:
That is true, but this option has been included in case support is added in the future.

uturuncoglu commented on ush/hafs_rtofs_download.py:
```diff
@@ -0,0 +1,230 @@
+#! /usr/bin/env python3
```
We are not supporting RTOFS at this point. Right?

evankalina replied:
Sam wrote some download scripts for it, but it is not supported in the workflow. Since he made the scripts, we can include them here in case RTOFS is supported by CDEPS in the future.

uturuncoglu reviewed (dismissed):
It looks fine to me

hyunsookkim-NOAA commented:
Hi Ufuk,
If this CDEP application pertains only to the regional HAFS application,
there is a way to support RTOFS. Let me know if you want to know about it.

Hyun-Sook

Hyun-Sook Kim
Physical Oceanography Division
Atlantic Oceanographic and Meteorological Laboratory
National Oceanic and Atmospheric Administration
4301 Rickenbacker Causeway
Miami, FL 33149


On Thu, Aug 19, 2021 at 4:13 PM Ufuk Turunçoğlu ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> It looks fine to me
> ------------------------------
>
> In parm/cdeps/datm_era5.streams
> <https://github.com/hafs-community/HAFS/pull/91#discussion_r692259555>:
>
> > @@ -0,0 +1,15 @@
> +stream_info:               ERA5_HOURLY01
> +taxmode01:                 limit
> +mapalgo01:                 bilinear
>
> We might also test mapalgo01: redist. It think there is no need to use
> bilinear in here since stream and datm component uses same grid/mesh.
> ------------------------------
>
> In parm/cdeps/docn_ghrsst.streams
> <https://github.com/hafs-community/HAFS/pull/91#discussion_r692259860>:
>
> > @@ -0,0 +1,15 @@
> +stream_info:               PRESCRIBED01
> +taxmode01:                 limit
> +mapalgo01:                 bilinear
>
> same also for this one, mapalgo01:redist
> ------------------------------
>
> In parm/cdeps/docn_oisst.streams
> <https://github.com/hafs-community/HAFS/pull/91#discussion_r692260195>:
>
> > @@ -0,0 +1,15 @@
> +stream_info:               PRESCRIBED01
> +taxmode01:                 limit
> +mapalgo01:                 bilinear
>
> same also for this one, mapalgo01:redist
> ------------------------------
>
> In parm/forecast/regional/nems.configure.cdeps.tmp
> <https://github.com/hafs-community/HAFS/pull/91#discussion_r692263526>:
>
> > +  MED med_phases_prep_ocn_avg
> +  MED -> ATM :remapMethod=redist
> +  MED -> OCN :remapMethod=redist
> +  ATM
> +  OCN
> +  MED med_phases_restart_write
> +  MED med_phases_history_write
> +@
> +::
> +
> +# Other Attributes #
> +DRIVER_attributes::
> +  start_type = startup
> +::
> +
> +ALLCOMP_attributes::
>
> Just for your information, the new version of CMEPS will also require
> setting following options under ALLCOMP_attributes::.
>
>   stop_n = [forecast length need to be defined in here and need to be consistent with the model_configure]
>   stop_option = nhours
>   stop_ymd = -999
>
> plus, new version of ESMF. So, once CMEPS is updated under HAFS (or UFS)
> ESMF module needs to be updated (8.2.0b?)
> ------------------------------
>
> In parm/hafs_basic.conf
> <https://github.com/hafs-community/HAFS/pull/91#discussion_r692266492>:
>
> > @@ -47,8 +47,15 @@ run_enkf=no          ;; Run self-cycled ensemble data assimilation system with E
>  run_wave=no          ;; Wave coupling (placeholder)
>  run_ocean=no         ;; Ocean coupling
>  ocean_model=hycom
> +run_datm=no          ;; Data atmosphere using CDEPS
> +run_docn=no          ;; Data ocean using CDEPS
> +run_dwav=no          ;; Data waves using CDEPS
>
> I think there is no data wave capability at this point. Right?
> ------------------------------
>
> In ush/hafs_rtofs_download.py
> <https://github.com/hafs-community/HAFS/pull/91#discussion_r692451559>:
>
> > @@ -0,0 +1,230 @@
> +#! /usr/bin/env python3
>
> We are not supporting RTOFS at this point. Right?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/hafs-community/HAFS/pull/91#pullrequestreview-734186122>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE736PDEXG64Q2H3BQDJLJTT5VQVXANCNFSM5CIDEHJQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

uturuncoglu commented:
@hyunsookkim-NOAA As i understand RTOFS data has issue in its grid and the mesh created using this grid has problem under ESMF. There might be option to use the RTOFS data by interpolating it to regular lat-lon or any other try-polar grid (I could able to use grid to create mesh for other tripolar grids that is used under CESM) externally (using tools like CDO etc.) that has no issue. Then, we could use those files to create CDEPS mode to support RTOFS but this involves extra regridding.

SamuelTrahanNOAA commented:
In a Grand Meeting about this PR, these changes were requested, and tasks were assigned:

1. [x] ~~Evan~~ Sam: Don't run the atmosphere initialization when running a data atmosphere model.
2. [x] Evan: Document how to run the download scripts directly.
3. [x] Sam: Move all cdeps ush stuff into ush/cdeps_utils
4. [x] Sam: Remove the docn/datm download jobs, and associated j-jobs and ex-scripts. Users will run the ush-level download scripts directly.
5. [x] Sam: Remove the docn/datm prep workflow jobs and J-jobs. Call their ex-scripts from the ocn/atm prep jobs instead.
6. [x] Sam: Remove Sam's python, and ensure the workflow runs without it when mesh generation is disabled.

evankalina commented:
@SamuelTrahanNOAA Thanks for this helpful summary. Unless you think otherwise, I won't do anything regarding item 1 yet, since it seems like it depends on item 5. Also, please let me know if you would like any help with testing.

SamuelTrahanNOAA committed:
Changes requested in review:
1. do not run the atmosphere initialization when running a data atmosphere model
2. Move all cdeps ush stuff into ush/cdeps_utils
3. Remove the docn/datm download jobs, and associated j-jobs and ex-scripts. Users will run the ush-level download scripts directly.
4. Remove the docn/datm prep workflow jobs and J-jobs. Call their ex-scripts from the ocn/atm prep jobs instead.
5. Remove Sam's python, and ensure the workflow runs without it when mesh generation is disabled.

SamuelTrahanNOAA committed:
Use batch for forecast job on Orion, to avoid overuse of "urgent"

SamuelTrahanNOAA commented:
All workflow changes requested in the Grand Meeting are complete. Evan still needs to update documentation. Also, someone needs to test this on Hera, Jet, and WCOSS, where Sam cannot for reasons not worth discussing on github.

evankalina commented:
And I have updated the documentation, which you can view [here](https://docs.google.com/document/d/1_UsLo96LAxCZjXHZe6qrr2WsR5W8PC3CGJxHod9qWKE). I will sync our branch with develop, make a few changes suggested by Ufuk, and run another test, this time on Jet/Hera. Then hopefully we can merge this in.

evankalina committed:
Merge branch 'develop' into feature/hafs_couplehycom_cdeps

evankalina committed:
Change mapalgo01 from bilinear to redist since the same grid is used
for the input data and data model (suggested by uturuncoglu).

evankalina committed:
Revert changes to build_forecast.sh.
We agreed to only specify -DAPP=HAFS in this file.
The user can change it to -DAPP=HAFS-ALL if they want
to build with CDEPS support, as described in the documentation.

evankalina commented:
@SamuelTrahanNOAA I'm currently testing DATM and DOCN on Hera and xjet. When running the initial sanity checks from run_hafs.py on Hera and xjet, the DATM run produces this error before submitting the launcher:

08/25/21 22:06:01 GMT :: hafs-hafs_couplehycom_cdeps-20210824_era5-00L-2019082900.xml :: Fatal error: Entity 'FORECAST_RESOURCES_regional_10x8io1x40_ocn120_omp1' not defined at :1.

Do you want me to modify the sites/hera.ent and sites/xjet.ent files to add this FORECAST_RESOURCES string, or does this point to a broader issue?

The DOCN runs have begun without incident.

SamuelTrahanNOAA commented:
Is it requesting the right number of ranks for each component? If so, please add the missing resources.

evankalina commented:
@SamuelTrahanNOAA Yes, the resource request matches the specifications in parm/hafs_datm_era5.conf. I added the missing resources and the runs are proceeding now. Thanks!

evankalina committed:
Small adjustments to job resources on Hera.

evankalina committed:
Small adjustment to resources on xjet.

evankalina commented:
@BinLiu-NOAA and others, all of the CDEPS forecasts for DATM (ERA5) and DOCN (GHRSST and OISST) succeeded on Hera. The DATM forecast succeeded on Jet, but the DOCN forecasts have been sitting in the queue for five days (they require more resources than DATM). At this point, maybe we should continue with this PR. Should we conduct the workflow-level RTs now?

BinLiu-NOAA commented on rocoto/cronjob_hafs_datm.sh:
```diff
@@ -0,0 +1,63 @@
+#!/bin/sh
```
@evankalina or @SamuelTrahanNOAA, could you please merge cronjob_hafs_datm.sh and cronjob_hafs_docn.sh into one file, saying cronjob_hafs_cdeps.sh? Thanks!

BinLiu-NOAA commented on rocoto/cronjob_hafs_datm.sh:
```diff
@@ -0,0 +1,63 @@
+     ../parm/hafs_regional_static.conf ../parm/hafs_hycom.conf \
+     ../parm/hafs_datm.conf ../parm/hafs_datm_era5.conf
+
+# Run all cycles of a storm
+#${PYTHON3} ./run_hafs.py ${dev} 2020 13L HISTORY config.EXPT=${EXPT} # Laura
+
+# Run specified cycles of a storm
+#${PYTHON3} ./run_hafs.py ${dev} 2020082506-2020082512 13L HISTORY \
+#   config.EXPT=${EXPT} config.SUBEXPT=${EXPT} # Laura
+
+# Run one cycle of a storm
+# ${PYTHON3} ./run_hafs.py -t ${dev} 2020082512 13L HISTORY config.EXPT=${EXPT}
+
```
The above commented out lines can probably be dropped.

BinLiu-NOAA commented on rocoto/hafs_workflow.xml.in:
```diff
@@ -228,7 +233,8 @@
     <dependency>
       <and>
-        <taskdep task="atm_prep"/>
+        <taskdep task="atm_prep"/> 
```
Please remove the whitespace at the end of this line.

BinLiu-NOAA commented on rocoto/hafs_workflow.xml.in:
```diff
@@ -277,21 +284,37 @@
           <!-- Or use COMgfs file dependency. Don't start until GFS 126h forecast is available.-->
           <or>
-@** if GFSVER==PROD2019
+@**   if GFSVER==PROD2019
```
Please revert the whitespace change.

BinLiu-NOAA commented on rocoto/hafs_workflow.xml.in:
```diff
@@ -277,21 +284,37 @@
             <datadep age="02:00" minsize="300000000"><cyclestr>&COMgfs;/gfs.@Y@m@d/@H/gfs.t@Hz.pgrb2.0p25.f126</cyclestr></datadep>
-@** endif
-@** if GFSVER==PROD2021
+@**   endif
+@**   if GFSVER==PROD2021
```
Please revert the whitespace change.

BinLiu-NOAA commented on rocoto/hafs_workflow.xml.in:
```diff
@@ -277,21 +284,37 @@
             <datadep age="02:00" minsize="300000000"><cyclestr>&COMgfs;/gfs.@Y@m@d/@H/atmos/gfs.t@Hz.pgrb2.0p25.f126</cyclestr></datadep>
             <datadep age="02:00" minsize="6000000000"><cyclestr>&COMgfs;/gfs.@Y@m@d/@H/atmos/gfs.t@Hz.atmf0126.nc</cyclestr></datadep>
-@** endif
+@**   endif
```
Please revert the whitespace change.

BinLiu-NOAA commented on rocoto/sites/kjet.ent:
```diff
@@ -25,7 +29,7 @@
   <!ENTITY ENKF_GSI_RESOURCES "<nodes>1:ppn=20:tpp=1</nodes><envar><name>TOTAL_TASKS</name><value>20</value></envar><envar><name>NCTSK</name><value>20</value></envar><envar><name>OMP_THREADS</name><value>1</value></envar><walltime>02:00:00</walltime>">
 
-  <!ENTITY FORECAST_WALLTIME "<walltime>07:59:00</walltime>">
+  <!ENTITY FORECAST_WALLTIME "<walltime>00:59:00</walltime>">
```
Please change the walltime limit back.

BinLiu-NOAA commented on sorc/build_forecast.sh:
```diff
@@ -6,6 +6,7 @@ cwd=`pwd`
 if [ $target = orion ]; then target=orion.intel ; fi
 if [ $target = jet ]; then target=jet.intel ; fi
+if [ $target = cheyenne ]; then target=cheyenne.intel ; fi
 
```
@evankalina, as we discussed, you might can use the following to replace the code section below:

```
if [ $target = wcoss_cray ]; then 
  app=HAFS
else
  app=HAFS-ALL
fi
cd hafs_forecast.fd/tests

./compile.sh "$target" "-DAPP=$app -DCCPP_SUITES=FV3_HAFS_v0_gfdlmp_tedmf_nonsst,FV3_HAFS_v0_gfdlmp_tedmf,FV3_HAFS_v0_hwrf_thompson,FV3_HAFS_v0_hwrf -D32BIT=ON" 32bit YES NO
```

In this case, one will not need to rebuild the forecast executable to run the HAFS CDEPS coupling configuration.

evankalina replied:
Great suggestion. I added this code and tested it for a data model run on Orion, and it is working.

BinLiu-NOAA commented:
> @BinLiu-NOAA and others, all of the CDEPS forecasts for DATM (ERA5) and DOCN (GHRSST and OISST) succeeded on Hera. The DATM forecast succeeded on Jet, but the DOCN forecasts have been sitting in the queue for five days (they require more resources than DATM). At this point, maybe we should continue with this PR. Should we conduct the workflow-level RTs now?

@evankalina I added a few review comments. I think once they are addressed, we are ready to run the HAFS application level RTs on HAFS supported platforms. We will also need @JohnSteffen-NOAA's help to run the HAFS CDEPS tests on wcoss_dell_p3. Thanks!

evankalina committed:
Changes requested from code review
* Delete unnecessary whitespace changes in hafs_workflow.xml.in
* Revert unintended change to forecast walltime on kjet
* Modify build_forecast.sh to build app=HAFS-ALL, except on wcoss_cray
* Combine the DATM and DOCN cronjob scripts into a general CDEPS
  cronjob, and delete commented out run commands.

evankalina committed:
Correct whitespace in CDEPS cron script.

evankalina commented:
@BinLiu-NOAA Thanks for the review! I believe I've addressed all of your comments in my latest commits. If you agree, we can start the regression testing.

BinLiu-NOAA commented:
> @BinLiu-NOAA Thanks for the review! I believe I've addressed all of your comments in my latest commits. If you agree, we can start the regression testing.

Thanks, @evankalina!

LinZhu-NOAA commented:
@BinLiu-NOAA In rocoto/hafs_multistorm_workflow.xml.in, the "@** endif" after the if condiction for RUN_OCEAN==YES at line 578 needs to move up to line 573 after if condition for RUN_GSI_VR==YES. Otherwise, it could not generate correct xml for globalnest tests.

BinLiu-NOAA commented:
> @BinLiu-NOAA In rocoto/hafs_multistorm_workflow.xml.in, the "@** endif" after the if condiction for RUN_OCEAN==YES at line 578 needs to move up to line 573 after if condition for RUN_GSI_VR==YES. Otherwise, it could not generate correct xml for globalnest tests.

Thanks, @LinZhu-NOAA! @evankalina or @SamuelTrahanNOAA, Could you please check and commit this change back into the rocoto/hafs_workflow.xml.in file of the feature/hafs_couplehycom_cdeps branch? Thanks!

evankalina committed:
Fix logic error in hafs_workflow.xml.in
* ocn_prep should be a dependency for the forecast job
  regardless of whether we are running the global nesting
  or regional HAFS. From Lin Zhu (IMSG at NOAA/EMC).

evankalina commented:
> > @BinLiu-NOAA In rocoto/hafs_multistorm_workflow.xml.in, the "@** endif" after the if condiction for RUN_OCEAN==YES at line 578 needs to move up to line 573 after if condition for RUN_GSI_VR==YES. Otherwise, it could not generate correct xml for globalnest tests.
> 
> Thanks, @LinZhu-NOAA! @evankalina or @SamuelTrahanNOAA, Could you please check and commit this change back into the rocoto/hafs_workflow.xml.in file of the feature/hafs_couplehycom_cdeps branch? Thanks!

Thanks @LinZhu-NOAA and @BinLiu-NOAA for noting this error! It has been fixed.

panll commented:
The regression tests finished smoothly with the latest commit on Hera. @evankalina @BinLiu-NOAA

LinZhu-NOAA commented:
The regression tests completed successfully on WCOSS-DELL.

mrinalbiswas commented:
@BinLiu-NOAA @evankalina RTs passed on Orion

BijuThomas-NOAA commented:
@BinLiu-NOAA @evankalina,
RTs were successful on WCOSS-Cray

BinLiu-NOAA commented:
Thanks for conducting the HAFS application level regression tests on hera, orion, wcoss_cray and wcoss_dell_p3!

@JohnSteffen-NOAA, have you get a chance to test the three HAFS CDEPS coupling test cases on wcoss_dell_p3 yet?

@evankalina, depending upon the status of @JohnSteffen-NOAA's HAFS CDEPS coupling tests on wcoss_dell_p3, we may decide to wait for his test results or to just go ahead merging this PR. In the meantime, in terms of how to merge this PR, not sure if you want choose the "squash and merge" approach or the regular merge approach. But in any case, it would be nice to include a comprehensive/detailed commit log message. Thanks!

JohnSteffen-NOAA commented:
@BinLiu-NOAA I have been working on the HAFS CDEPS tests for wcoss_dell_p3 as of yesterday and will continue today. I will keep you updated.

BinLiu-NOAA commented:
> @BinLiu-NOAA I have been working on the HAFS CDEPS tests for wcoss_dell_p3 as of yesterday and will continue today. I will keep you updated.

@JohnSteffen-NOAA, Thanks for keeping us updated with your HAFS CDEPS coupling tests on wcoss_dell_p3!

evankalina commented:
Thanks @BijuThomas-NOAA, @LinZhu-NOAA, @panll, and @mrinalbiswas for taking the time to conduct these regression tests, and to @JohnSteffen-NOAA for testing HAFS CDEPS on wcoss_dell_p3!

@BinLiu-NOAA I think this PR is a good candidate for the "squash and merge" approach. We made so many individual commits (85!), so a regular merge will make the commit history look noisy. I can write a clean commit message summarizing all of the changes instead when I squash merge it. I will wait a bit to hear from @JohnSteffen-NOAA first.

BinLiu-NOAA commented:
> Thanks @BijuThomas-NOAA, @LinZhu-NOAA, @panll, and @mrinalbiswas for taking the time to conduct these regression tests, and to @JohnSteffen-NOAA for testing HAFS CDEPS on wcoss_dell_p3!
> 
> @BinLiu-NOAA I think this PR is a good candidate for the "squash and merge" approach. We made so many individual commits (85!), so a regular merge will make the commit history look noisy. I can write a clean commit message summarizing all of the changes instead when I squash merge it. I will wait a bit to hear from @JohnSteffen-NOAA first.

Agreed. The squash and merge method sounds perfectly good to me as well. Thanks @evankalina!

JohnSteffen-NOAA committed:
Adds 2 ENTITY FORECAST_RESOURCES lines for running HAFS-CDEPS on wcoss_dell_p3 and adds CDO module load to modulefile

BinLiu-NOAA reviewed (approved):
HAFS application regression tests passed. And HAFS CDEPS coupling tests also ran through on Hera/Orion/Wcoss_dell_p3.
